### PR TITLE
Fix modal run on class method defined on base of wrapped class

### PR DESCRIPTION
**Pull request by mwaskom**
Fixes CLI-22

Hard to explain exactly what the issue was here: it's fairly complex code. But there was a gap in the conditionals for resolving the name of the method to execute given (1) a Python function object (i.e. an unbound method) and (2) a Python class that has been registered to an App.

We have a lot of undefined behavior and edge cases around class inheritance.

## Changelog

- Fixed a bug where pointing `modal run` at a method on a Modal Cls would fail if the method was inherited from a parent.

### CODE REVIEW

File: test/supports/class_hierarchy.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+class Base:
+    @modal.method()
+    def defined_on_base(self):
+        print("base")
+
+    @modal.method()
+    def overridden_on_wrapped(self):
+        raise NotImplementedError()
+
```
Comment (mwaskom) [inline test/supports/class_hierarchy.py:14]: Was hoping this would also fix the bug where `modal run my_app.py::Wrapped.overridden_on_wrapped` executes the base version of the method, but apparently that is a separate defect.

Comment (mwaskom) [reply test/supports/class_hierarchy.py:15]: Resolved here: https://github.com/modal-labs/modal-client/pull/2590

**Comment by mwaskom** [comment]
@prbot approve; unblocks a user

File: modal/_utils/function_utils.py [hunk 2]
```diff
@@ -133,6 +133,9 @@ def __init__(
         elif f is None and user_cls:
             # "service function" for running all methods of a class
             self.function_name = f"{user_cls.__name__}.*"
+        elif f and user_cls:
```
Comment (freider) [inline modal/_utils/function_utils.py:136]: I don't think we should even be constructing FunctionInfo for methods (except @build ones, and we might want to deprecate those), but that's a separate matter.

**Review by freider — COMMENTED** [review]
nice catch! should get rid of Methods even using FunctionInfo (which should really be ServiceInfo)
